services/federator/test/unit/Test/Federator/Monitor.hs

```diff
@@ -395,15 +397,26 @@ testMergeWatchedPaths =
          in mergedPaths == origPaths
     ]
 
-newtype Path = Path {getPath :: FilePath}
+newtype Path = Path {getRawPath :: ByteString}
+
+getPath :: Path -> IO FilePath
+getPath = fromRawPath . getRawPath
+
+getAbsolutePath :: Path -> IO FilePath
+getAbsolutePath p = do
+  path <- getPath p
+  makeAbsolute ("/" <> path)
 
 instance Show Path where
-  show = show . getPath
+  show = show . getRawPath
 
 instance Arbitrary Path where
-  arbitrary = Path . intercalate "/" <$> listOf (listOf1 ch)
+  arbitrary =
+    Path . B8.intercalate "/"
+      <$> listOf (BS.pack <$> listOf1 ch)
     where
-      ch = arbitrary `suchThat` (/= '/')
+      ch :: Gen Word8
+      ch = arbitrary `suchThat` (/= fromIntegral (ord '/'))
```

Comment:
Probably need to exclude `0` / ASCII `NUL`, too right?

Reply:
I'm not 100% sure, but I don't think so. It seems the roundtripping encoding used by GHC can handle arbitrary bytes. A 0 byte should just become a zero character, which is a valid code point, and roundtrip back to a 0 byte.

Reply:
?  Then, what does Char(s)? does `0x30 :: Word8` map to and vice-versa?  Also, whatever GHC roundtripping is happening should handle `0x2F` / `'/'`, but we still exclude it...

If this is going through some other form of filtering to make sure the Characters are valid in path components, why do we need to exclude something that maps to `/`?  If this is the only such filtering / processing, then we need to exclude stuff that outputs ASCII `NUL`.

It didn't look like there was any additional processing before this was used to access to the filesystem, though...

Reply:
> `0x30 :: Word8` is mapped to unicode code point `\u0030`, i.e. `0`. Not sure why that would be special.

Then, shouldn't `0 :: Word8` be mapped to `'\u0000' :: Char` (aka ASCII NUL), which isn't a valid character in a file path, under most encodings.

Reply:
> > Isn't that what I wrote two comments ago?
> 
> You wrote "A 0 byte should just become a zero character" which I interpreted as `0 :: Word8` maps to `'0' :: Char`. 

Oh, I see the confusion now. Sorry about that.

> I'm fine with leaving it as is if the test stays reliable. I think the filtering of `'/'` paired with the `useAsCString` just had me in the mindset that we needed the byte string to be a valid path component on Linux / POSIX, as `'/'` and NUL are the two banned bytes. (`':'` is also problematic in places, because of PATH-like things.)

Alright, thanks. Let's leave it for now then. I'll still wait for confirmation from @jschaul before merging, since I think he's the only one who's actually witnessed a failure so far.